### ISSUE TITLE
Bind caddy on a Unix Domain Socket UDS with http/2 only

### ISSUE THREAD

**Issue by Zetanova** [caddyserver/caddy#2999]
I am trying with the docker image caddy/caddy:alpine to host a reverse proxy
from a unix domain socket to an unix domain socket.

The whole purpose is to overwrite the http2 ":authority" header with "localhost".
header ":authority: /tmp/caddy.sock"  should be header ":authority: localhost"

the connection is comming from an gRPC golang client (h2, no tls). 
this does not support ALPN and need to be in http/2 

Test Caddyfile
```
http://localhost

bind unix//tmp/caddy.sock

reverse_proxy unix//tmp/upstream.sock {
    header_up :authority localhost
}
```

Questions / Bugs:
-- How to create a catch to any hostname or a section with an socket name like "/tmp/caddy.sock"? 
-- How to force to http2 ? 
-- If caddy is binding to an UDS, it will create a file but including the port "/tmp/caddy.sock:80"

**Comment by mholt** [comment]
Thanks for the questions! This is something I want to get working in Caddy 2 if it doesn't already work, but I won't be able to look at it for about a week... can someone else take a look?

**Comment by mholt** [comment]
> -- How to create a catch to any hostname or a section with an socket name like "/tmp/caddy.sock"?

I don't know what this means, can you rephrase?

> -- How to force to http2 ?

Can't be done over plain HTTP; but can be done using the TLS directive. Set the ALPN values so that only `h2` is present.

> -- If caddy is binding to an UDS, it will create a file but including the port "/tmp/caddy.sock:80"

Fixed! Thanks.

**Comment by Zetanova** [comment]
caddy can bind to UDS but the config section requires a hostname in http.
There is no option to overwrite an h2 pseudoheader ":authority" to the upstream proxy

```
#instead of http://localhost or ANY
unix//tmp/caddy.sock 

bind unix//tmp/caddy.sock

reverse_proxy unix//tmp/upstream.sock {
    header_up :authority localhost
}
```

I would need a reverse proxy to overwrite the  :authority pseudo header
Kubernetes Plugin => gRPC over UDS => Caddy => Kestrel

**Comment by Zetanova** [comment]
@mholt 

h2 has no host header anymore it is now ":authority" pseudoheader.
The request from kubernetes is coming over UDS without network and the host aka ":authority" header has the unix-path in it and not a http-url. 
In caddy config there is no option to make a section for it, because it requires a valid http-url for the route. 

To overwrite the ":authority" header field for the upstream request is the whole purpose for me.
golang itself can handle h2 RFC fine, but i am very poor in golang itself.

nginx and apache is validating the h2 ":authority" header too
and assuming that it needs to be a valid http-url.

dotnet Kestrel issue:
https://github.com/dotnet/aspnetcore/issues/18522#issuecomment-600846993
 
I would need caddy or a simple golang proxy as a workaround for my kubernetes CSI plugin:
https://github.com/Zetanova/hyperv-csi-driver

**Comment by mholt** [comment]
> h2 has no host header anymore it is now ":authority" pseudoheader.

Yes, I know that. 🙂 But Go takes care of the H1/H2 translations under the hood of the standard lib, hence why I suggest using `Host`. Did you at least try it?

**Comment by Zetanova** [comment]
@mholt i can try it after i can resolve the issue with the config section

i could not find a method to make a section for a unix-path route or at least a catch-all.
in nginx this would be the "default server"

**Comment by mholt** [comment]
Try just using `http://` or `:80` then.

**Comment by Zetanova** [comment]
@mholt Ok, i will try it, but both should logicly not match.

**Comment by Zetanova** [comment]
I am testing with the 'caddy/caddy' docker image

How to force http2?
gRPC need to run over http2 without negation/upgrade.

The bug with the port in the socket got resolved.

Caddyfile:
```
http://

bind unix//csi/caddy.sock

reverse_proxy unix//csi/csi.sock {
    header_up Host localhost
    #transport grpc
}
```

Output:
```
"level":"info","ts":1585793125.678197,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":"caddyfile"}
{"level":"info","ts":1585793125.693246,"logger":"admin","msg":"admin endpoint started","address":"localhost:2019","enforce_origin":false,"origins":["localhost:2019"]}  
2020/04/02 02:05:25 [INFO][cache:0xc000614230] Started certificate maintenance routine
{"level":"info","ts":1585793125.6935596,"logger":"tls","msg":"cleaned up storage units"}
{"level":"info","ts":1585793125.693799,"msg":"autosaved config","file":"/config/caddy/autosave.json"}
{"level":"info","ts":1585793125.6938102,"msg":"serving initial configuration"}
{"level":"error","ts":1585793126.2471795,"logger":"http.log.error","msg":"net/http: HTTP/1.x transport connection broken: malformed HTTP response \"\\x00\\x00\\b\\a\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x01\"","request":{"method":"PRI","uri":"*","proto":"HTTP/2.0","remote_addr":"@","host":"","headers":{}},"status":502,"err_id":"4vv1nu9pc","err_trace":"reverseproxy.(*Handler).ServeHTTP (reverseproxy.go:363)"}
{"level":"error","ts":1585793126.3967094,"logger":"http.log.error","msg":"net/http: HTTP/1.x transport connection broken: malformed HTTP response \"\\x00\\x00\\b\\a\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x01\"","request":{"method":"PRI","uri":"*","proto":"HTTP/2.0","remote_addr":"@","host":"","headers":{}},"status":502,"err_id":"b5ezz6czk","err_trace":"reverseproxy.(*Handler).ServeHTTP (reverseproxy.go:363)"}
```

**Comment by mholt** [comment]
@Zetanova Thanks for trying it! Please ask on [our forums](https://caddy.community) -- as that seems to be a different, unrelated question.

**Comment by Zetanova** [comment]
@mholt I write a post under https://caddy.community/t/v2-force-http2-unencrypted-on-an-uds-path-bind/7398
